File: pilot/src/api/chaos.rs
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::IntoResponse;
use axum::Json;
use std::collections::HashMap;
use serde::{Deserialize, Serialize};

use crate::app_state::AppState;
use crate::chaos::types::{ChaosAction, ChaosScenario};

use super::error::ApiError;

// --- Request / Response types ---

#[derive(Debug, Deserialize)]
pub struct CreateScenarioRequest {
    pub name: String,
    #[serde(default = "default_namespace")]
    pub namespace: String,
    pub actions: Vec<ChaosAction>,
}

fn default_namespace() -> String {
    "regret-system".to_string()
}

#[derive(Debug, Serialize)]
pub struct ScenarioResponse {
    pub id: String,
    pub name: String,
    pub namespace: String,
    pub actions: Vec<ChaosAction>,
    pub created_at: String,
}

#[derive(Debug, Serialize)]
pub struct InjectionResponse {
    pub id: String,
    pub scenario_id: String,
    pub scenario_name: String,
    pub status: String,
    pub started_at: String,
    pub finished_at: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct StartInjectionResponse {
    pub injection_id: String,
    pub scenario_id: String,
    pub status: String,
}

// --- Scenario CRUD ---

pub async fn create_scenario(
    State(state): State<AppState>,
    Json(req): Json<CreateScenarioRequest>,
) -> Result<impl IntoResponse, ApiError> {
    if req.actions.is_empty() {
        return Err(ApiError::BadRequest("actions must not be empty".to_string()));
    }

    let id = format!("cs-{}", uuid::Uuid::now_v7());
    let actions_json = serde_json::to_string(&req.actions)
        .map_err(|e| ApiError::BadRequest(e.to_string()))?;

    let record = state
        .sqlite
        .create_chaos_scenario(&id, &req.name, &req.namespace, &actions_json)
        .await?;

    Ok((
        StatusCode::CREATED,
        Json(scenario_to_response(&record)),
    ))
}

pub async fn list_scenarios(
    State(state): State<AppState>,
) -> Result<impl IntoResponse, ApiError> {
    let records = state.sqlite.list_chaos_scenarios().await?;
    let items: Vec<ScenarioResponse> = records.iter().map(scenario_to_response).collect();
    Ok(Json(serde_json::json!({ "items": items })))
}

pub async fn get_scenario(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let record = state
        .sqlite
        .get_chaos_scenario(&id)
        .await?
        .ok_or_else(|| ApiError::NotFound(format!("chaos scenario {id} not found")))?;
    Ok(Json(scenario_to_response(&record)))
}

pub async fn update_scenario(
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(req): Json<CreateScenarioRequest>,
) -> Result<impl IntoResponse, ApiError> {
    state
        .sqlite
        .get_chaos_scenario(&id)
        .await?
        .ok_or_else(|| ApiError::NotFound(format!("chaos scenario {id} not found")))?;

    if req.actions.is_empty() {
        return Err(ApiError::BadRequest("actions must not be empty".to_string()));
    }

    let actions_json = serde_json::to_string(&req.actions)
        .map_err(|e| ApiError::BadRequest(e.to_string()))?;

    let record = state
        .sqlite
        .update_chaos_scenario(&id, &req.name, &req.namespace, &actions_json)
        .await?
        .ok_or_else(|| ApiError::Internal(anyhow::anyhow!("update failed")))?;

    Ok(Json(scenario_to_response(&record)))
}

pub async fn delete_scenario(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let deleted = state.sqlite.delete_chaos_scenario(&id).await?;
    if !deleted {
        return Err(ApiError::NotFound(format!("chaos scenario {id} not found")));
    }
    Ok(StatusCode::NO_CONTENT)
}

// --- Injection control ---

#[derive(Debug, Deserialize, Default)]
pub struct InjectRequest {
    /// Override action params by action type. e.g. {"upgrade_test": {"candidate_image": "..."}}
    #[serde(default)]
    pub overrides: HashMap<String, serde_json::Value>,
}

pub async fn start_injection(
    State(state): State<AppState>,
    Path(scenario_id): Path<String>,
    body: Option<Json<InjectRequest>>,
) -> Result<impl IntoResponse, ApiError> {
    let record = state
        .sqlite
        .get_chaos_scenario(&scenario_id)
        .await?
        .ok_or_else(|| ApiError::NotFound(format!("chaos scenario {scenario_id} not found")))?;

    let mut actions: Vec<ChaosAction> = serde_json::from_str(&record.actions)
        .map_err(|e| ApiError::Internal(anyhow::anyhow!("invalid actions JSON: {e}")))?;

    // Apply overrides — merge override params into matching action types
    if let Some(Json(req)) = body {
        for action in &mut actions {
            if let Some(overrides) = req.overrides.get(&action.action_type) {
                if let Some(override_obj) = overrides.as_object() {
                    if let Some(params_obj) = action.params.as_object_mut() {
                        for (k, v) in override_obj {
                            params_obj.insert(k.clone(), v.clone());
                        }
                    }
                }
            }
        }
    }

    let scenario = ChaosScenario {
        id: record.id.clone(),
        name: record.name.clone(),
        namespace: record.namespace.clone(),
        actions,
    };

    let injection_id = state.chaos.start_injection(scenario).await?;

    Ok((
        StatusCode::ACCEPTED,
        Json(StartInjectionResponse {
            injection_id,
            scenario_id: record.id,
            status: "running".to_string(),
        }),
    ))
}

pub async fn stop_injection(
    State(state): State<AppState>,
    Path(injection_id): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    state.chaos.stop_injection(&injection_id).await?;
    Ok(StatusCode::NO_CONTENT)
}

pub async fn list_injections(
    State(state): State<AppState>,
) -> Result<impl IntoResponse, ApiError> {
    let records = state.sqlite.list_chaos_injections().await?;
    let items: Vec<InjectionResponse> = records
        .iter()
        .map(|r| InjectionResponse {
            id: r.id.clone(),
            scenario_id: r.scenario_id.clone(),
            scenario_name: r.scenario_name.clone(),
            status: r.status.clone(),
            started_at: r.started_at.clone(),
            finished_at: r.finished_at.clone(),
            error: r.error.clone(),
        })
        .collect();
    Ok(Json(serde_json::json!({ "items": items })))
}

pub async fn get_injection(
    State(state): State<AppState>,
    Path(injection_id): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let record = state
        .sqlite
        .get_chaos_injection(&injection_id)
        .await?
        .ok_or_else(|| ApiError::NotFound(format!("chaos injection {injection_id} not found")))?;
    Ok(Json(InjectionResponse {
        id: record.id,
        scenario_id: record.scenario_id,
        scenario_name: record.scenario_name,
        status: record.status,
        started_at: record.started_at,
        finished_at: record.finished_at,
        error: record.error,
    }))
}

pub async fn delete_injection(
    State(state): State<AppState>,
    Path(injection_id): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    // Stop if still running
    if state.chaos.is_active(&injection_id).await {
        state.chaos.stop_injection(&injection_id).await?;
    }
    let deleted = state.sqlite.delete_chaos_injection(&injection_id).await?;
    if !deleted {
        return Err(ApiError::NotFound(format!(
            "chaos injection {injection_id} not found"
        )));
    }
    Ok(StatusCode::NO_CONTENT)
}

// --- Chaos events ---

pub async fn chaos_events(
    State(state): State<AppState>,
) -> Result<impl IntoResponse, ApiError> {
    let events = state.files.read_chaos_events(None)?;
    let mut body = String::new();
    for event in &events {
        body.push_str(&serde_json::to_string(event).unwrap_or_default());
        body.push('\n');
    }
    Ok((
        [(axum::http::header::CONTENT_TYPE, "application/x-ndjson")],
        body,
    ))
}

// --- Helpers ---

fn scenario_to_response(
    r: &crate::storage::sqlite::ChaosScenarioRecord,
) -> ScenarioResponse {
    let actions: Vec<ChaosAction> =
        serde_json::from_str(&r.actions).unwrap_or_default();
    ScenarioResponse {
        id: r.id.clone(),
        name: r.name.clone(),
        namespace: r.namespace.clone(),
        actions,
        created_at: r.created_at.clone(),
    }
}

File: pilot/endpoints/chaos.go
package endpoints

import (
	"github.com/regret-io/regret/pilot-go/ext"
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/go-chi/chi/v5"

	chaospkg "github.com/regret-io/regret/pilot-go/chaos"
	"github.com/regret-io/regret/pilot-go/database"
)

type chaosHandlers struct {
	state *AppState
}

// CreateScenario handles POST /api/chaos/scenarios.
func (c *chaosHandlers) CreateScenario(w http.ResponseWriter, r *http.Request) {
	var req CreateScenarioRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		WriteError(w, BadRequest("invalid JSON: "+err.Error()))
		return
	}

	if req.Namespace == "" {
		req.Namespace = "regret-system"
	}
	if len(req.Actions) == 0 {
		WriteError(w, BadRequest("actions must not be empty"))
		return
	}

	id := fmt.Sprintf("cs-%d", ext.NowUnixMilli())
	actionsJSON, _ := json.Marshal(req.Actions)

	record, err := c.state.Sqlite.CreateChaosScenario(r.Context(), id, req.Name, req.Namespace, string(actionsJSON))
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	WriteJSON(w, http.StatusCreated, scenarioToResponse(record))
}

// ListScenarios handles GET /api/chaos/scenarios.
func (c *chaosHandlers) ListScenarios(w http.ResponseWriter, r *http.Request) {
	records, err := c.state.Sqlite.ListChaosScenarios(r.Context())
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}
	var items []ScenarioResponse
	for i := range records {
		items = append(items, scenarioToResponse(&records[i]))
	}
	WriteJSON(w, http.StatusOK, map[string]interface{}{"items": items})
}

// GetScenario handles GET /api/chaos/scenarios/{id}.
func (c *chaosHandlers) GetScenario(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	record, err := c.state.Sqlite.GetChaosScenario(r.Context(), id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("chaos scenario %s not found", id)))
		return
	}
	WriteJSON(w, http.StatusOK, scenarioToResponse(record))
}

// UpdateScenario handles PUT /api/chaos/scenarios/{id}.
func (c *chaosHandlers) UpdateScenario(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	ctx := r.Context()

	_, err := c.state.Sqlite.GetChaosScenario(ctx, id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("chaos scenario %s not found", id)))
		return
	}

	var req CreateScenarioRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		WriteError(w, BadRequest("invalid JSON: "+err.Error()))
		return
	}

	if len(req.Actions) == 0 {
		WriteError(w, BadRequest("actions must not be empty"))
		return
	}

	actionsJSON, _ := json.Marshal(req.Actions)

	if err := c.state.Sqlite.UpdateChaosScenario(ctx, id, req.Name, req.Namespace, string(actionsJSON)); err != nil {
		WriteError(w, InternalError(err))
		return
	}

	record, err := c.state.Sqlite.GetChaosScenario(ctx, id)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}
	WriteJSON(w, http.StatusOK, scenarioToResponse(record))
}

// DeleteScenario handles DELETE /api/chaos/scenarios/{id}.
func (c *chaosHandlers) DeleteScenario(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	deleted, err := c.state.Sqlite.DeleteChaosScenario(r.Context(), id)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}
	if !deleted {
		WriteError(w, NotFound(fmt.Sprintf("chaos scenario %s not found", id)))
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// StartInjection handles POST /api/chaos/scenarios/{id}/inject.
func (c *chaosHandlers) StartInjection(w http.ResponseWriter, r *http.Request) {
	scenarioID := chi.URLParam(r, "id")
	ctx := r.Context()

	record, err := c.state.Sqlite.GetChaosScenario(ctx, scenarioID)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("chaos scenario %s not found", scenarioID)))
		return
	}

	var actions []chaospkg.ChaosAction
	if err := json.Unmarshal([]byte(record.Actions), &actions); err != nil {
		WriteError(w, InternalError(fmt.Errorf("invalid actions JSON: %w", err)))
		return
	}

	// Apply overrides if provided
	var req InjectRequest
	if r.Body != nil {
		_ = json.NewDecoder(r.Body).Decode(&req)
	}
	if len(req.Overrides) > 0 {
		for i := range actions {
			if override, ok := req.Overrides[actions[i].ActionType]; ok {
				var overrideObj map[string]interface{}
				if json.Unmarshal(override, &overrideObj) == nil {
					if actions[i].Params == nil {
						actions[i].Params = make(map[string]interface{})
					}
					for k, v := range overrideObj {
						actions[i].Params[k] = v
					}
				}
			}
		}
	}

	scenario := chaospkg.ChaosScenario{
		ID:        record.ID,
		Name:      record.Name,
		Namespace: record.Namespace,
		Actions:   actions,
	}

	injectionID, err := c.state.Chaos.StartInjection(ctx, scenario)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	WriteJSON(w, http.StatusAccepted, StartInjectionResponse{
		InjectionID: injectionID,
		ScenarioID:  record.ID,
		Status:      "running",
	})
}

// StopInjection handles POST /api/chaos/injections/{id}/stop.
func (c *chaosHandlers) StopInjection(w http.ResponseWriter, r *http.Request) {
	injectionID := chi.URLParam(r, "id")
	if err := c.state.Chaos.StopInjection(r.Context(), injectionID); err != nil {
		WriteError(w, InternalError(err))
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// ListInjections handles GET /api/chaos/injections.
func (c *chaosHandlers) ListInjections(w http.ResponseWriter, r *http.Request) {
	records, err := c.state.Sqlite.ListChaosInjections(r.Context())
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}
	var items []InjectionResponse
	for _, rec := range records {
		items = append(items, InjectionResponse{
			ID:           rec.ID,
			ScenarioID:   rec.ScenarioID,
			ScenarioName: rec.ScenarioName,
			Status:       rec.Status,
			StartedAt:    rec.StartedAt,
			FinishedAt:   rec.FinishedAt,
			Error:        rec.Error,
		})
	}
	WriteJSON(w, http.StatusOK, map[string]interface{}{"items": items})
}

// GetInjection handles GET /api/chaos/injections/{id}.
func (c *chaosHandlers) GetInjection(w http.ResponseWriter, r *http.Request) {
	injectionID := chi.URLParam(r, "id")
	record, err := c.state.Sqlite.GetChaosInjection(r.Context(), injectionID)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("chaos injection %s not found", injectionID)))
		return
	}
	WriteJSON(w, http.StatusOK, InjectionResponse{
		ID:           record.ID,
		ScenarioID:   record.ScenarioID,
		ScenarioName: record.ScenarioName,
		Status:       record.Status,
		StartedAt:    record.StartedAt,
		FinishedAt:   record.FinishedAt,
		Error:        record.Error,
	})
}

// DeleteInjection handles DELETE /api/chaos/injections/{id}.
func (c *chaosHandlers) DeleteInjection(w http.ResponseWriter, r *http.Request) {
	injectionID := chi.URLParam(r, "id")
	ctx := r.Context()

	// Stop if still running
	if c.state.Chaos.IsActive(injectionID) {
		_ = c.state.Chaos.StopInjection(ctx, injectionID)
	}

	deleted, err := c.state.Sqlite.DeleteChaosInjection(ctx, injectionID)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}
	if !deleted {
		WriteError(w, NotFound(fmt.Sprintf("chaos injection %s not found", injectionID)))
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// ChaosEvents handles GET /api/chaos/events.
func (c *chaosHandlers) ChaosEvents(w http.ResponseWriter, r *http.Request) {
	events, err := c.state.Files.ReadChaosEvents(nil)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	w.Header().Set("Content-Type", "application/x-ndjson")
	w.WriteHeader(http.StatusOK)
	for _, event := range events {
		data, _ := json.Marshal(event)
		w.Write(data)
		w.Write([]byte("\n"))
	}
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func scenarioToResponse(r *database.ChaosScenarioRecord) ScenarioResponse {
	var actions []json.RawMessage
	_ = json.Unmarshal([]byte(r.Actions), &actions)
	return ScenarioResponse{
		ID:        r.ID,
		Name:      r.Name,
		Namespace: r.Namespace,
		Actions:   actions,
		CreatedAt: r.CreatedAt,
	}
}
